marker_api/src/ast/item.rs

```diff
@@ -238,42 +240,182 @@ use impl_item_data;
+        self.span.copy().map(|span| with_cx(self, |cx| cx.span(span)))
     }
-}
 
-#[cfg(feature = "driver-api")]
-impl<'ast> Visibility<'ast> {
-    pub fn new(item_id: ItemId) -> Self {
-        Self {
-            _lifetime: PhantomData,
-            _item_id: item_id,
+    /// Returns `true` if the item is declared as public, without any restrictions.
+    ///
+    /// ```
+    /// // This returns `true`
+    /// pub fn unicorn() {}
+    ///
+    /// // This returns `false`, since the visibility is restricted to a specified path.
+    /// pub(crate) fn giraffe() {}
+    ///
+    /// // This returns `false`, since the visibility is not defined
+    /// fn dragon() {}
+    /// ```
+    ///
+    /// See [`Visibility::is_pub_with_path`] to detect pub declarations with a
+    /// defined path.
+    pub fn is_pub(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Public | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns `true` if the item is declared as `pub(...)` with a path in brackets
+    /// that defines the scope, where the item is visible.
+    pub fn is_pub_with_path(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Path(_) | VisibilityKind::Crate(_))
+    }
+
+    /// Returns `true` if the visibility is declared as `pub(crate)`. This is a
+    /// special case of the `pub(<path>)` visibility.
+    ///
+    /// This function checks if the visibility is restricted and the defined path
+    /// belongs to the root module of the crate. Meaning, that this can also be `true`,
+    /// if the visibility uses `pub(super)` to travel up to the crate root.
+    ///
+    /// ```rs
+    /// // lib.rs
+    ///
+    /// mod example_1 {
+    ///     // Returns `false` since no visibility is declared
+    ///     fn foo() {}
+    ///
+    ///     // Returns `false` since the item is not visible from the root of the crate.
+    ///     pub(in crate::example_1) fn bar() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to the root of the crate.
+    ///     pub(crate) baz() {}
+    ///
+    ///     // Returns `true` as the visibility is restricted to `super` which is the
+    ///     // root of the crate.
+    ///     pub(crate) boo() {}
+    /// }
+    ///
+    /// // Returns `false` since the visibility is not restricted
+    /// fn example_in_root() {}
+    /// ```
+    pub fn is_pub_crate(&self) -> bool {
+        matches!(self.kind, VisibilityKind::Crate(_))
+    }
+
+    /// Returns `true` if a visibility has been defined.
+    pub fn is_defined(&self) -> bool {
+        !matches!(self.kind, VisibilityKind::Default(_) | VisibilityKind::DefaultPub)
+    }
+
+    /// Returns the [`ItemId`] of the module where this item is visible in, if the
+    /// visibility is defined to be public inside a specified path.
+    ///
+    /// See [`Visibility::module_id`] to get the `ItemId`, even if the item or
+    /// field uses the default visibility.
+    pub fn pub_with_path_module_id(&self) -> Option<ItemId> {
```

Comment:
This again looks weird due to the mixture of syntactic and semantic properties of the visibility.

I think if we split the representation into `ast` and `sem` there would be only a method on `sem` type:

- `sem::Visibility::visibility_root() -> Option<ItemId>`: returns the `ItemId` of the root module this visibility level allows access in; returns `None` if the visibility is `pub`

And there wouldn't be such a method for `ast::Visibility`. I think it would be more straightforward to just expose the syntactic `VisibilityKind` to them, and the users would then handle all possible syntactic flavors of the visibility they want to check for.
But the `VisibilityKind` in the `ast` module would look like this:

```rust
enum VisibilityKind {
  Pub,
  PubIn(Path),
  PubCrate,
  
  /// There is no distinction between `Default/DefautPub` because in the synatctic representation
  /// we only care if the visibility was explicitly specified or not
  Default, 
}
```

Maybe I'm dreaming here, and rustc doesn't provide the original syntax form of the visibility, in which case it would be a bummer.

**UPD**: I noticed the `module_id` method only after writing this comment.